services/heartbeat/etcd/store.go

```diff
@@ -294,6 +306,47 @@ func servicePrefix(service string) string {
 	return fmt.Sprintf(keyFormat, heartbeatKeyPrefix, service)
 }
 
-func leaseKey(service, instance string, ttl time.Duration) string {
-	return fmt.Sprintf(keyFormat, heartbeatKey(service, instance), ttl.String())
+func newLeaseCache() *leaseCache {
+	return &leaseCache{
+		leases: make(map[string]map[time.Duration]clientv3.LeaseID),
+	}
+}
+
+type leaseCache struct {
+	sync.RWMutex
+
+	leases map[string]map[time.Duration]clientv3.LeaseID
+}
+
+func (c *leaseCache) get(service, instance string, ttl time.Duration) (clientv3.LeaseID, bool) {
+	c.RLock()
+	leases, ok := c.leases[heartbeatKey(service, instance)]
```

Comment:
Might be better to just defer `c.RUnlock()`  so you don't need to track where the function might return